### PR TITLE
Boot time RNG seeding.

### PR DESCRIPTION
**Pull request by paulidale**
Add a mechanism that ensures that /dev/urandom is properly seeded before
drawing any material from it in a manner that prevent starvation or blocking
of processes after the initial seeding has been completed.

### CODE REVIEW

**Comment by paulidale** [comment]
This PR is a bit rough and ready, but the idea behind it should be clear enough to promote some discussion either here or on the project list.

Configuration time conditionals and filesystem paths could be desirable.

**Comment by t8m** [comment]
Certainly the directory path where the seeded file is placed needs to be configurable and for the whole functionality it should be possible to be disabled through the build configuration.

**Comment by vdukhovni** [comment]
The use of fixed file names in `/tmp` almost invariably comes with various security exposures and DoS opportunities. This is no exception. A hostile actor can create a mode `000` file in /tmp that nobody else can write to, or create a symlink to a critical file where writing a zero byte might do some damage. Or lie and get applications to continue without adequate entropy.

If this is to be a **trusted** signal, it needs to be set by a trusted agent. On modern Linux systems in some file under `/run` that only root can write, with a suitable boot-time service creating the file early (existence should be enough, no need to write anything there). The service in question would need to be a dependency of all other services that need secure seeding.

But, if only the trusted service can write the file, then there's no need in libcrypto to have code to write the file, since the trusted service can do that directly without libcrypto.  But then there's also nothing one can do if the file is not there, a correctly configured system will have already seeded the RNG.  Which means that there no need for any code in libcrypto, since there's nothing it can do.

If any process can unwedge the system-wide CSPRNG, we have potential security risks, if only root can do it at boot, then nobody else can or should do anything after that.

**Comment by mspncp** [comment]
There's nothing to add to @vdukhovni's analysis. I completely agree.

**Comment by mspncp** [comment]
I think Victor's analysis shows once more that the problem of initial seeding of the CSPRNG can only be solved properly by the kernel and/or the init system, not by ordinary user programs.

File: crypto/rand/rand_unix.c [hunk 1]
```diff
@@ -325,6 +325,44 @@ static struct random_device {
...
+        if (st.st_size == 0) {
+            rfd = open(random_path, O_RDONLY);
+            if (rfd == -1 || read(rfd, &c, 1) != 1 || write(fd, &c, 1) != 1)
+                goto err;
+            close(rfd);
+            close(fd);
+            seeded = 1;
+        }
+    }
+    return;
```
Comment (levitte) [inline crypto/rand/rand_unix.c]: I wonder, if two processes using libcrypto are started at the same time, and one of them creates that file and waits for /dev/random to respond, what happens with the other process?  Does it wait for the first process to be done with this part, or does it simply fail?

Comment (paulidale) [reply crypto/rand/rand_unix.c]: The `flock(fd, LOCK_EX)` takes care of this, other processes wait.

**Comment by levitte** [comment]
> Certainly the directory path where the seeded file is placed needs to be configurable and for the whole functionality it should be possible to be disabled through the build configuration.

`/tmp` is usually emptied at boot.  Other directories are not, which would mean that the file would linger across reboots, making the code in this PR utterly useless (we do *not* want to go through the hoops involved in checking it this is a fresh boot or not).

(I know, `/tmp` isn't guaranteed to be emptied, it's just what usually happens as long as the sysadmin doesn't set something different up...  and that's another fragility in this PR)

**Comment by paulidale** [comment]
Agreed, _solving_ this is the job of the operating system.  Mitigating (i.e. _lessening_) the risk is something that we reasonably should be able to do.

The problem is low entropy during and soon after startup.  What assumptions can be made?  Is it reasonable to assume that libcrypto is loaded during the boot sequence or can it be delayed an arbitrarily long time?

What is the threat model?

- An attacker who has control during the boot sequence?  This is required to create the file before libcrypto has had a chance to load.
- A privileged attacker after the boot sequence?  This is required to remove the file from a sticky directory.
- Something else entirely?

The first doesn't seem all that credible, the second isn't gaining capabilities they wouldn't already have (overwrite files, `ln /dev/zero /dev/urandom`).


I'm happy to abandon this path if it is believed it cannot be made workable.  I'm also open to other suggestions as to what we can do.

File: crypto/rand/rand_unix.c (hunk 1 above)
Comment (levitte) [reply crypto/rand/rand_unix.c]: Ah, okie.

**Comment by paulidale** [comment]
> Other directories are not, which would mean that the file would linger across reboots, making the code in this PR utterly useless (we do _not_ want to go through the hoops involved in checking it this is a fresh boot or not).

I'm not sure there is a way to determine if there has been a fresh boot reliably.
It might be possible to use a file containing material read from a properly seeded RND and to persist this across reboots -- it is more involved and less performant.  Every instance of libcrypto would need to open the file, get a write lock, read the contents to seed it's RNG, generate new data to replace the contents of the file and then unlock and close.  Ideally it would want to overwrite the contents on exit too.

It isn't a path I'd be keen on following but it might be possible.

**Comment by levitte** [comment]
> I'm not sure there is a way to determine if there has been a fresh boot reliably.

"reliably" is a relative term, but there's always the possibility to look for the latest `BOOT_TIME` record in wtmp and compare its timestamp with the timestamp of your file.

> It might be possible to use a file containing material read from a properly seeded RND and to persist this across reboots -- it is more involved and less performant. Every instance of libcrypto would need to open the file, get a write lock, read the contents to seed it's RNG, generate new data to replace the contents of the file and then unlock and close. Ideally it would want to overwrite the contents on exit too.

Are we looking at the return of the `.rnd` file?  That's basically what you're describing, although perhaps (I don't quite remember) with better quality.

**Comment by t8m** [comment]
/var/run is also usually emptied. And as the early boot processes are usually run as root it does not matter that regular user cannot write there. But anyway the path should be build-time configurable. (Regardless of whether this approach really makes sense or not.)

**Comment by richsalz** [comment]
TL;DR: This is bad.  Either you wait for the random device to be ready, or you do nothing.

**Comment by petrovr** [comment]
Pauli wrote:
> Add a mechanism that ensures that /dev/urandom is properly seeded before
> drawing any material from it in a manner that prevent starvation or blocking
> of processes after the initial seeding has been completed.

If I remember well all linux-es seed urandom on boot/reboot time . File 
/{etc|var/lib}/random-seed is created by "shutdown" script.

Dunno why OpenSSL needs to duplicate existing OS-functionality.

Regards,
Roumen

**Comment by kroeckx** [comment]
Any solution based on a file does not work in a chroot by default. So the system administrator would at least do something so that it also works in the chroot. One alternative that has been suggested is to use shared memory with some known key. That could also work in a chroot.

To be secure, both solution would need to be created by root and be read only for the rest. I guess it also means that we actually need to check the owner (and permissions). I think this requires that the init system is responsible for creating something we can wait on.

Since the init system needs to be modified, you could argue that the init system should also prevent services from being started that want to use /dev/urandom, so that we don't actually ever need to wait. It would just be a signal that it's an init system that provides us with a secure /dev/urandom. But not everything is a service, people might try to run some command that needs /dev/urandom. So I think that in any case the init system needs to confirm to us that /dev/urandom has been initialized.

**Comment by vdukhovni** [comment]
> @paulidale The problem is low entropy during and soon after startup. What assumptions can be made? Is it reasonable to assume that libcrypto is loaded during the boot sequence or can it be delayed an arbitrarily long time?
Not all systems will do crypto at boot. So it could be delayed, unless the init system provides a service that guarantees CSPRNG startup at boot, with a built-in wait for adequate seeding.

> What is the threat model?

With "trusted" files in /tmp? ... Anything goes.

> I'm happy to abandon this path if it is believed it cannot be made workable. I'm also open to other suggestions as to what we can do.

The best I can see happening along this path, is that the files in question would have to be owned by root in /run or /var/run (which are empty on reboot on Linux systems). The platform would then provide a boot service that initializes the file in question at boot.  Libcrypto code when running as root, and perhaps only within a configurable time after boot as determined via sysinfo(2), could run your logic to conditionally wait for `/dev/random` to be ready, if no other root process has already done that.

This eliminates DoS via locks on the /tmp file, hostile symlinks, mode 000 files, etc.  Is it a good idea? I am not convinced one way or the other as yet (even though I'm suggesting it).

**Comment by paulidale** [comment]
> With "trusted" files in /tmp? ... Anything goes.

I'm aware of a number of problems /tmp has had over the years, they don't seem relevant here.

An attacker that has control of the system at boot time is a concern.  Accepting this risk makes the mode 000, symbolic links and DoS via write locking or creating the file early moot.

After boot, the file is in a sticky directory, has its permissions restricted and isn't locked anymore -- root would be required to do anything nasty unless I'm missing something.

An installation (i.e. first start) after boot could suffer from a DoS via a write lock and would risk symbolic link evil.  An uptime check might help.  This is also the chroot case (if the chroot happens before seeding).

**Comment by bernd-edlinger** [comment]
I think we depend too much on the system RNG source to be error-free.
Different O/S do have rather different quality of the RNG data they offer.
How I see the issue, is that prior to 1.1.1 there were a lot of .rnd files all over
which are now no longer used, but on the other hand, if the random
source is maybe not so reliable, then accumulating random data from
previous runs will make a difference.
So maybe we depend now more on the correctness of the random data
sources than before 1.1.1 since we do no longer use the .rnd files?

**Comment by bernd-edlinger** [comment]
I think we depend too much on the system RNG source to be error-free.
Different O/S do have rather different quality of the RNG data they offer.
How I see the issue, is that prior to 1.1.1 there were a lot of .rnd files all over
which are now no longer used, but on the other hand, if the random
source is maybe not so reliable, then accumulating random data from
previous runs will make a difference.
So maybe we depend now more on the correctness of the random data
sources than before 1.1.1 since we do no longer use the .rnd files?

**Comment by paulidale** [comment]
In many ways, I agree with this viewpoint.  Others aren't so accomodating.

I do have another idea for "just after boot" entropy collection.  I'll need time to explore things and it quite possibly will come to nothing.  Failing this, we've got a dilemma for the 3.0.0 release which needs to address this underlying concern.

**Comment by paulidale** [comment]
Close this one because of the unrequited love it garnered.

**Comment by mspncp** [comment]
> Close this one because of the unrequited love it garnered.

Let me guess: A citation from Dr. Who or from the Hitchhiker's Guide? Or is it even from Shakespeare? ;-)

**Comment by paulidale** [comment]
None of the above :)
Just me feeling inquisitioned.
